Tests for Uri.php

daweeb opened the pull request:
Many functions have a Throw exception listed in phpdocu, but no test to trigger it.
Is this Wanted? @midahp
https://github.com/daweeb/Http/blob/06ff3334b051c556e2bd297cc7ee63efb20b0d0e/src/Uri.php#L363

https://github.com/daweeb/Http/blob/06ff3334b051c556e2bd297cc7ee63efb20b0d0e/src/Uri.php#L422

Done
-  Host
-  Port
- Scheme
- Fragment
- UserInfo
- remove comments
-  csfixer/phpformat
-  authority

Commit:
added tests for Port Host and Fragment

Commit:
added scheme userinfo and Fragment Test

Commit:
auth + nullStandard test added, formatted + phpfixer

Commit:
add invalidPort Error

Commit:
added test in Uri.php

midahp commented on test/UriTest.php:
```diff
@@ -52,4 +55,155 @@ public function testToString()
+        $this->assertEquals($uri->getPort(), $port);
+    }
+
+    public function testWithInvalidPort()
```
Add a test for port <0. Then add a check in Uri.php.

daweeb replied:
done

midahp commented on test/UriTest.php:
```diff
@@ -52,4 +55,155 @@ public function testToString()
+        $port = '';
+        $uri = new Uri();
+        $uri = $uri->withPort($port);
+        $this->assertEquals($uri->getPort(), null);
```
This is a good example where run into problems with `assertEquals`. This will succeed if `$uri->getPort()` returns 0 (which is what happens here). Use `assertSame` to check the type aswell.
In this case you should use `assertNull($uri->getPort())` though, because you already know you want a `null` to be returned.

daweeb replied:
done, but hat to add a new if in withport function. There is probably a smarter way of doing this. I will add a link when pushed to repo. https://github.com/daweeb/Http/blob/942b837266d1043954177cb91d2c0ad1d75636c8/src/Uri.php#L383

midahp commented on test/UriTest.php:
```diff
@@ -52,4 +55,155 @@ public function testToString()
+        $host = 'group@ware';
+        $uri = new Uri();
+        $uri = $uri->withHost($host);
+        $this->assertEquals($uri->getHost(), $host);
```
If you want to add this check, this test should expect an exception for this value. Then add that exception in Uri.php

daweeb replied:
I added a hostname check, please validate if correct use-case applies here ( FILTER_VALIDATE_URL  would be better here, i guess).  Again added a null check.

midahp commented on test/UriTest.php:
```diff
@@ -52,4 +55,155 @@ public function testToString()
+        $host = '';
+        $uri = new Uri();
+        $uri = $uri->withHost($host);
+        $this->assertEquals($uri->getHost(), null);
```
You should use assertSame('', $uri->getHost());

daweeb replied:
done

midahp commented on test/UriTest.php:
```diff
@@ -52,4 +55,155 @@ public function testToString()
+        $this->assertEquals($uri->getHost(), null);
+    }
+
+    public function testWithPathRemovesFragmentString()
```
This test already exists above.

daweeb replied:
removed

midahp commented on test/UriTest.php:
```diff
@@ -52,4 +55,155 @@ public function testToString()
+        $fragment = '';
+        $uri = new Uri();
+        $uri = $uri->withFragment($fragment);
+        $this->assertEquals($uri->getFragment(), null);
```
use `assertSame('', $uri->getFragment()`

daweeb replied:
done

midahp commented on test/UriTest.php:
```diff
@@ -52,4 +55,155 @@ public function testToString()
+        $uri = new Uri();
+        $scheme = '';
+        $uri = $uri->withScheme($scheme);
+        $this->assertEquals($uri->getScheme(), null);
```
use `assertSame('', $uri->getScheme())`

daweeb replied:
done

midahp commented on test/UriTest.php:
```diff
@@ -52,4 +55,155 @@ public function testToString()
+        $uri = $uri->withHost($host);
+        $uri = $uri->withUserInfo($user);
+        $uri = $uri->withPort($port);
+        $this->assertEquals($uri->getAuthority(), '');
```
use `assertSame` here

daweeb replied:
done

midahp reviewed (changes requested):
Please solve the mentioned problems and try to add the necessary checks in Uri.php.
Also try to add tests for the `with*` methods to make sure they don't modify the original object

Commit:
added checks and tests for with(Host/Port)

Commit:
added Test for functions with clone

midahp commented on src/Uri.php:
```diff
@@ -339,6 +340,17 @@ public function withUserInfo($user, $password = null): self
      */
     public function withHost($host): self
     {
+        if ($host == null) {
```
Use `if (!$host)` instead

daweeb replied:
done

midahp commented on src/Uri.php:
```diff
@@ -339,6 +340,17 @@ public function withUserInfo($user, $password = null): self
+        }
+
+        if (!filter_var($host, FILTER_VALIDATE_DOMAIN, FILTER_FLAG_HOSTNAME)) {
+            throw new InvalidArgumentException('Invalid Hostname, Valid characters for hostnames are ASCII(7) letters.');
```
This message is not correct right? Not all ASCII letters are valid.

daweeb replied:
changed, sry.

midahp commented on src/Uri.php:
```diff
@@ -364,6 +376,17 @@ public function withHost($host): self
     public function withPort($port): self
     {
+        if ($port >= 65536 || $port <0) {
+            throw new InvalidArgumentException('Invalid Portnumber, port must be between 1 and 65535');
```
Port 0 is valid.

daweeb replied:
changed. sry.

midahp commented on src/Uri.php:
```diff
@@ -364,6 +376,17 @@ public function withHost($host): self
+            throw new InvalidArgumentException('Invalid Portnumber, port must be between 1 and 65535');
+        }
+
+        if ($port == null) {
```
use:
```
if (!is_null($port) {
    $port = (int) $port;
}
$ret = clone $this;
$ret->port = $port;
return $ret;
```

daweeb replied:
done, with error
`1) Horde\Http\Test\UriTest::testWithNullPort Failed asserting that 0 is null`
is this wanted? In the PHPDoc it says 
`  A null value provided for the port is equivalent to removing the port information.`
so i thought it should be null not 0.

daweeb replied:
From Ralf:

0 is an integer and an invalid port number - you must error.
null is "no port number given" and valid.

Check for null first and skip the port range check if it is null.

changed it.

midahp replied:
Yes I mentioned it in another comment. But my code suggestion should still work. You test just tested the wrong thing.

midahp commented on test/UriTest.php:
```diff
@@ -52,4 +55,154 @@ public function testToString()
         $this->assertEquals($url, (string) $uri);
     }
+
+    public function testWithValidPort()
```
It would probably be useful to test the edge cases with port = 0 and port = 65535

daweeb replied:
good idea, done.

midahp commented on test/UriTest.php:
```diff
@@ -52,4 +55,154 @@ public function testToString()
+        $this->assertEquals($uri->getHost(), 'groupware');
+    }
+
+    public function testWithInvalidHost()
```
Add a test with invalid host that only has ascii characters. Like `test##Host`.
Add tests for valid hosts that use the `-` and the `.` characters.
Add a test for invalid host that starts with the `-` character.

daweeb replied:
Added.

midahp commented on test/UriTest.php:
```diff
@@ -52,4 +55,154 @@ public function testToString()
+        $uri = $uri->withHost($host);
+    }
+
+    public function testWithNullHost()
```
I would call this `testWithEmptyHost` but its fine

daweeb replied:
changed.

midahp reviewed (changes requested):
Added some comments.
Also try to add tests for the with* methods to make sure they don't modify the original object

ralflang commented:
You are confusing things.

0 is an integer and an invalid port number - you must error.
null is "no port number given" and valid.

Check for null first and skip the port range check if it is null.

Commit:
modified testcases for UriTest.php and functions for Uri.php

modified Exception comments in Uri.php for With(Port/Host)
modified withPort if port range statement in Uri.php
modified withHost if host is null statement in Uri.php
added valid/invalid host tests in UriTest.php
added With methods tests when clone in UriTest.php

Commit:
Modified test name for understanding

midahp commented on src/Uri.php:
```diff
@@ -364,6 +376,17 @@ public function withHost($host): self
+            return $ret;
+        }
+
+        if ($port >= 65536 || $port <= 0) {
```
I think you missunderstood. I meant the Exception message needs to  say `between 0 and 65535`. You actually made port 0 invalid in the if. Port 0 is a valid port (as far as I know).

midahp replied:
My mistake.

midahp commented on test/UriTest.php:
```diff
@@ -52,4 +58,259 @@ public function testToString()
+        $this->assertEquals($uri->getPort(), $port);
+    }
+
+    public function testWithValidPortMin()
```
This needs to be adjusted, because port 0 is a valid port. See other comment about `withPort`
